### PR TITLE
Makes kube-proxy flags take precedence on proxy config

### LINKED ISSUE #98302: kube-proxy flags are not respected when using config file together
1. With following config file and start kube-proxy with flags `--config=/tmp/kube-proxy.yaml --master=https://localhost:6443 --profiling=true`. 
```
cat <<EOF > /tmp/kube-proxy.yaml
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
clientConnection:
  kubeconfig: /var/run/kubernetes/kube-proxy.kubeconfig
hostnameOverride: 127.0.0.1
mode:
featureGates:
  AllAlpha: false
EOF
```

2. Try to access the debug endpoint, you'll get 404
```
curl 127.0.0.1:10249/debug/pprof/
404 page not found

```

Actually, not only the `profiling` flag does not take effect, all other flags which are not set in config file, are all changed to default value and then do not take effect.

### PR DESCRIPTION
**Pull request by hanlins**
Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Kube-proxy flags are silently ignored when `--config` is set. This PR makes flags take precedence on the proxy component configuration.

#### Which issue(s) this PR fixes:

Fixes #98302 

#### Does this PR introduce a user-facing change?

```release-note
Make kube-proxy flags take precedence on kube-proxy component configuration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kube-proxy/app/server_test.go [hunk 1]
```diff
@@ -625,3 +625,76 @@ func TestAddressFromDeprecatedFlags(t *testing.T) {
 
 	}
 }
+
+func TestFlagOverwrites(t *testing.T) {
```
Comment (knabben) [inline cmd/kube-proxy/app/server_test.go:629]: Don't need to be on this PR but we can extend the testing for platform and instances specific ones.

**Comment by jayunit100** [comment]
@thockin this is the thing you were looking for, `mergo`, good job discovering it @hanlins !

File: cmd/kube-proxy/app/server_test.go (hunk 1 above)
Comment (hanlins) [reply cmd/kube-proxy/app/server_test.go:629]: Thanks @knabben, good point! I'll add coverage once reviewers reach a consensus on this overwriting behavior.

Comment (hanlins) [reply cmd/kube-proxy/app/server_test.go:629]: Thanks @knabben, good point! I'll add coverage once reviewers reach a consensus on this overwriting behavior. The test for now is just to demo flags will take precedence on proxy component configs.

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -228,6 +229,8 @@ func (o *Options) Complete() error {
 		if err != nil {
 			return err
 		}
+		// make flags take precedence
+		mergo.Merge(c, o.config, mergo.WithOverride)
```
Comment (liggitt) [inline cmd/kube-proxy/app/server.go:233]: does this make defaulted values override explicit config values?

**Comment by hanlins** [comment]
> @thockin this is the thing you were looking for, `mergo`, good job discovering it @hanlins !

Credits to @aojea!

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (hanlins) [reply cmd/kube-proxy/app/server.go:233]: No, have a test case here: https://github.com/kubernetes/kubernetes/blob/2e112f2efe15907af9b2dc2e485e74bc980df429/cmd/kube-proxy/app/server_test.go#L663-L668
You can see if the flag is not set but proxy component config is set, after the merge the value will be the one in proxy component config.

Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: the unit test that makes sure this works doesn't match how Options are actually used by kube-proxy

kube-proxy:
1. creates options: `opts := NewOptions()`
2. applies defaults: `opts.config, err = opts.ApplyDefaults(opts.config)`
3. binds flags

I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

In the unit test, make sure you are exercising this the same way, and make sure the config file is effective when setting a value that is defaulted by opts.ApplyDefaults

**Comment by liggitt** [comment]
This takes kube-proxy in a different direction than other components (kube-scheduler explicitly makes config and flags exclusive, with the exception of a few deprecated flags).

Whether or not flags should override config generally has been long-debated (cc @mtaufen @deads2k @sttts). I am of the opinion they should not, since it makes inspection/validation of a config file in isolation difficult to understand without also having the full invocation it is a part of.

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: > I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

If that turns out to be the case, that means our testing of the config file path has gaps that need to be closed as well

Comment (hanlins) [reply cmd/kube-proxy/app/server.go:233]: I see, that's a good point. I think the tricky part is it's hard to determine whether the value is from the default or explicitly passed from flags. The merge logic here is to make the values in `o.config` take precedence if it's a non-empty value, which can be our defaults.
I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: > I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

That makes the help displayed for the flags much less helpful, since it doesn't show the defaults used if no flags are passed. Before restructuring defaulting, we should determine whether this change is the right direction, see the comment in https://github.com/kubernetes/kubernetes/pull/101273#issuecomment-823357862

**Comment by hanlins** [comment]
Hi @liggitt, there're lots of back and forth on this flag precedence behavior, and I agree with your points. But one key difference between components like `kube-scheduler` and `kube-proxy` is `kube-proxy` sometimes needs instance-specific configurations, so in this sense, it's more similar to `kubelet` rather than `kube-scheduler`. Originally we plan to just use `--config` and `--instance-config` to cover both cases, but as you know https://github.com/kubernetes/enhancements/pull/1439 has been out for a while, but it still has some gaps to fill and could take a few more releases to get there. 
Before that instance-specific config happens, the only thing user have for per-instance config is the flags, but most of the flags are just silently ignored, it's like we want to provide people a better semantics for configurations in the future, we gave them one which is not ideal and the one we gave is not working. Last time during the WG meeting @thockin mentioned that we want to make our existing flag users happy, and we want to do flag overwriting similar to `kubelet` if the cost is not too large. So we take this direction on fixing the flag overwriting for our existing users.

**Comment by liggitt** [comment]
> we want to do flag overwriting similar to `kubelet`

Having reviewed the kubelet flag/config merging changes originally, and many fixup changes over the years, I would strongly recommend against that path. The kubelet config initialization path is almost incomprehensible.

**Comment by jayunit100** [comment]
hmm... so, @liggitt are you suggesting we eliminate these flags entirely ?

**Comment by liggitt** [comment]
"fail if config + flags overlap" avoids adding complex new merge behavior, and avoids the footgun of ignored flags. I'd start with that.

**Comment by hanlins** [comment]
Hi @liggitt @knabben, made some updates here. I have to add some struct tag to associate the fields in `KubeProxyConfiguration` with the flags people use. With the change, we can detect and fast fail if the user sets the field in kube-proxy config and also sets the flag for it. But, to address the issue where people's flags got ignored silently, we still need to aggregate the configs from the config file and flags to one place, but since we've already checked that there's no conflict between config file and flags, we can safely combine the two without worry about the precedence. Any opinion on that?

**Comment by liggitt** [comment]
> With the change, we can detect and fast fail if the user sets the field in kube-proxy config and also sets the flag for it.

What's the benefit to letting them set via config file field and flag but only when they match?

How does this determine the difference between a default value in the config and an explicitly set value in the config?

File: staging/src/k8s.io/component-base/config/types.go [hunk 3]
```diff
@@ -23,17 +23,17 @@ import (
 // ClientConnectionConfiguration contains details for constructing a client.
 type ClientConnectionConfiguration struct {
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/config/types.go:24]: This type is shared between multiple components (including out-of-tree components), so these flags are determined by the consuming component.

Comment (hanlins) [reply staging/src/k8s.io/component-base/config/types.go:24]: I see. I checked other components like kubelet, and they used the same flag for the same field, but I wasn't aware that maybe some external components are also consuming them. Would it be better if I mark make the struct tag something more specific to kube-proxy? Like change the tag from `flag` to `kubeProxyFlag`?

**Comment by hanlins** [comment]
> What's the benefit to letting them set via config file field and flag but only when they match?
> 
> How does this determine the difference between a default value in the config and an explicitly set value in the config?

Actually, if the config is set in both the config file and the flag, no matter if they have the same value or not, we will consider this as a conflict and fail the command. In the test I have I'm using different values but actually it doesn't matter.

In our detection logic, we're actually detecting if the flag value is explicitly set by the user, or it's something from the flag default https://github.com/kubernetes/kubernetes/blob/9a9960d886dac65fc0cba17b99d50c3e546a5779/cmd/kube-proxy/app/server.go#L478
Basically, it's using this [Changed](https://pkg.go.dev/github.com/spf13/pflag#FlagSet.Changed) interface to detect whether the flag is set by user or not.

**Comment by mtaufen** [comment]
> Whether or not flags should override config generally has been long-debated (cc @mtaufen @deads2k @sttts). I am of the opinion they should not, since it makes inspection/validation of a config file in isolation difficult to understand without also having the full invocation it is a part of.

I'm in agreement with this. We did it in kubelet because we were pursuing an incremental migration across releases and were trying to keep things backwards compatible when a flag later became available in config, but might still be set on the command line in an existing configuration. But if you can keep it either-or, the implementation is much simpler.
